Tests branch merge to main

owhere opened the pull request:
### Summary
This PR implements following features
- show test coverage using codecov 
- increases test coverage from 53% to 96%.
- interprets new JWT from OIDC

### Related Task
Resolves Issue #7 New projects format from JWT
Resolves Issue #3 Deserialization of projects claim
Work for [Project #76](https://github.com/isambard-sc/jupyter-platform/issues/76)
Work for [Project #68](https://github.com/isambard-sc/jupyter-platform/issues/68)

Commit:
adding tests coverage

Commit:
bump version of upload-artifact

Commit:
install test coverage for github action

Commit:
generage coverage html reports

Commit:
Only upload coverage report on python 3.12

Commit:
fix indentation

Commit:
show codecov icon

Commit:
generate xml report for codecov

Commit:
add codecov token

Commit:
show icon on tests branch

Commit:
add tests for login handler

Commit:
linting fixes

Commit:
add config for asyncio

Commit:
ignore codecov for now

Commit:
tests on authenticator and handler get

Commit:
update imports

Commit:
upload to codecov

Commit:
save coverage xml as artefact

Commit:
upgrade upload artifact

Commit:
only generate coverage for python3.1.2

Commit:
unique name for coverage report

Commit:
use coverage config

Commit:
install pytest-asyncio

Commit:
remove coverage config

Commit:
refactor and test spawner

Commit:
linting

Commit:
set codecov threshold

Commit:
update badge colour

Commit:
test requested user and homedir

Commit:
test project options

Commit:
update projects from new JWT

Commit:
script to decode a JWT token

Commit:
linting

Commit:
test projects

Commit:
test old and new format of projects

Commit:
get projects for spawner to authorize

jcwomack committed:
Add `pytest-asyncio` and `pytest-cov` to dev dependencies

jcwomack committed:
Explicitly configure asyncio mode and loop scope

Configuring loop scope suppresses a DeprecationWarning when running tests

jcwomack committed:
Use TEST_DIR make variable

jcwomack commented on codecov.yml:
```diff
@@ -0,0 +1,8 @@
+        target: 85  
+  color: 
+    green: 85  
+    yellow: 70  
```
```suggestion
```

The `coverage.color` field is apparently invalid when I try to validate the `codecov.yml` file (see <https://docs.codecov.com/docs/codecov-yaml#validate-your-repository-yaml>).

```text
jw5533 in bricsauthenticator at 2025-02-11T14:59:11
> cat codecov.yml 
coverage:
  status:
    project:
      default:
        target: 85  
  color: 
    green: 85  
    yellow: 70  
                                                                                                                                                             
jw5533 in bricsauthenticator at 2025-02-11T14:59:18
> curl --data-binary @codecov.yml https://codecov.io/validate
Error at ['coverage', 'color']: 
unknown field
```

jcwomack commented on README.md:
```diff
@@ -1,5 +1,7 @@
 # bricsauthenticator
 
+[![codecov](https://codecov.io/gh/owhere/bricsauthenticator/branch/tests/graph/badge.svg?token=3IGKBYAIS1)](https://codecov.io/gh/owhere/bricsauthenticator)
```
I think this badge might be showing the coverage for `owhere/bricsauthenticator`, not `isambard-sc/bricsauthenticator`.

It looks like the upload to Codecov is failing in Actions, see e.g. <https://github.com/isambard-sc/bricsauthenticator/actions/runs/13262845648/job/37023173498>

```text
 [2025-02-11T12:25:43.623Z] ['error'] There was an error running the uploader: Error uploading to https://codecov.io: Error: There was an error fetching the storage URL during POST: 429 - {"message":"Rate limit reached. Please upload with the Codecov repository upload token to resolve issue. Expected time to availability: 1914s."}

[2025-02-11T12:25:43.623Z] ['info'] Codecov will exit with status code 0. If you are expecting a non-zero exit code, please pass in the `-Z` flag
```

jcwomack replied:
I've made an issue to track this: https://github.com/isambard-sc/bricsauthenticator/issues/10

jcwomack commented on .github/workflows/build_lint_test.yml:
```diff
@@ -2,9 +2,9 @@ name: "Python package: build, lint, and test"
 on:
   push:
-    branches: [ "main" ]
+    branches: [ "main", "tests" ]
```
```suggestion
    branches: [ "main" ]
```

Suggest we default to only running automatically on the `main` branch to avoid excessive use of GitHub Actions minutes.

The `on.workflow_dispatch:` key allows the workflow to be manually run on any branch from the Actions tab on GitHub, e.g. for build, lint, test workflow, go to <https://github.com/isambard-sc/bricsauthenticator/actions/workflows/build_lint_test.yml> then select the "Run workflow" drop-down menu.

jcwomack commented on .github/workflows/build_lint_test.yml:
```diff
@@ -2,9 +2,9 @@ name: "Python package: build, lint, and test"
+    branches: [ "main", "tests" ]
   pull_request:
-    branches: [ "main" ]
+    branches: [ "main", "tests" ]
```
```suggestion
    branches: [ "main" ]
```

Suggest to use manual workflow dispatch when targeting branches other than `main`

jcwomack committed:
Use env var to store version of Python to run coverage for

jcwomack committed:
Delete debugging step in GitHub Actions workflow

jcwomack commented on src/bricsauthenticator/auth.py:
```diff
@@ -13,46 +13,63 @@
+        signing_key = self._fetch_signing_key(jwks_uri, token)
+        decoded_token = self._decode_jwt(token, signing_key, signing_algos)
+
+        # Example usage of _normalize_projects:
```
```suggestion
```

jcwomack commented on src/bricsauthenticator/spawner.py:
```diff
@@ -40,7 +51,8 @@ def auth_state_hook(spawner, auth_state) -> None:
             if auth_state:
                 spawner.log.debug(f"Acquired auth_state: {str(auth_state)}")
-                spawner.brics_projects = auth_state
+                projects = {k.split(".")[0]: v for k, v in auth_state.items()}
+                spawner.brics_projects = projects
```
```suggestion
                spawner.brics_projects = {k.split(".")[0]: v for k, v in auth_state.items()}
```

jcwomack committed:
Move script to new scripts/ directory

jcwomack committed:
Add scripts dir to format and lint make targets

jcwomack commented on tests/test_auth.py:
```diff
@@ -0,0 +1,262 @@
+    handler._parse_oidc_config = MagicMock(return_value=(["RS256"], "https://example.com/jwks"))
+    handler._fetch_signing_key = MagicMock(return_value=MagicMock(key="mock_key"))
+    handler._decode_jwt = MagicMock(return_value={"short_name": "test_user", "projects": {}})
+    handler._normalize_projects = MagicMock(return_value={"project1": "value1"})
```
I think the return value of this function should be a dict of lists

```suggestion
    handler._normalize_projects = MagicMock(return_value={"project1": ["value1"]})
```

jcwomack commented on tests/test_auth.py:
```diff
@@ -0,0 +1,262 @@
+    handler._fetch_signing_key = MagicMock(return_value=MagicMock(key="mock_key"))
+    handler._decode_jwt = MagicMock(return_value={"short_name": "test_user", "projects": {}})
+    handler._normalize_projects = MagicMock(return_value={"project1": "value1"})
+    handler.auth_to_user = AsyncMock(return_value={"name": "test_user"})
```
This isn't the right return type for `auth_to_user()`. I think it is a `User` object, see <https://github.com/jupyterhub/jupyterhub/blob/718f01e600ba55a65618d820161b72073767a3f8/jupyterhub/handlers/base.py#L899>


```suggestion
		user = MagicMock()
		user.name = "test_user"
    handler.auth_to_user = AsyncMock(return_value=user)
```

jcwomack commented on tests/test_auth.py:
```diff
@@ -0,0 +1,262 @@
+    )
+    handler._fetch_signing_key.assert_called_once_with("https://example.com/jwks", "mock_token")
+    handler._decode_jwt.assert_called_once_with("mock_token", handler._fetch_signing_key.return_value, ["RS256"])
+    handler.set_login_cookie.assert_called_once_with({"name": "test_user"})
```
```suggestion
    handler.set_login_cookie.assert_called_once_with(user)
```

See above!

jcwomack commented on tests/test_spawner.py:
```diff
@@ -0,0 +1,90 @@
+    spawner.user.name = "testuser"
+    spawner.user_options = {"brics_project": "project1"}
+
+    # Call the method
+    result = spawner._req_username_default()
```
```suggestion
    # Call the method (which is a traitlets dynamic default value for the req_username trait)
    result = spawner.req_username
```

jcwomack commented on tests/test_spawner.py:
```diff
@@ -0,0 +1,90 @@
+    spawner.user.name = "testuser"
+    spawner.user_options = {"brics_project": "project1"}
+
+    # Call the method
+    result = spawner._req_homedir_default()
```
```suggestion
    # Call the method (which is a traitlets dynamic default value for the req_username trait)
    result = spawner.req_homedir
```

jcwomack reviewed (commented):
Many thanks Fan for your work on this! I really like how you have modularised the code and used mocking for testing.

jcwomack committed:
Apply suggestions from code review

Only automatically run workflows on pushes to `main` and PRs with `main` as target

Workflows can be manually run on other branches/commits using `workflow_dispatch`

jcwomack committed:
Update codecov.yml

Remove invalid codecov.yml keys

jcwomack committed:
Update src/bricsauthenticator/auth.py

Remove outdated comment

jcwomack committed:
Update src/bricsauthenticator/spawner.py

Remove unnecessary variable assignment

jcwomack committed:
Update tests/test_auth.py

Fix mocked _normalize_projects() return value

jcwomack committed:
Update tests/test_auth.py

Use mock object as more realistic return value for mocked `auth_to_user()` method

jcwomack committed:
Update tests/test_auth.py

Use mock object as more realistic return value for mocked `auth_to_user()` method

jcwomack committed:
Apply suggestions from code review

Check result of traitlet directly rather than calling function used as dynamic default

jcwomack committed:
Remove stray tabs

jcwomack reviewed (approved):
I am happy for this to be merged. Once it is merged I will tag a new minor version in `main`.

My only minor concerns are 

* It doesn't seem like the coverage data is being uploaded to Codecov in the Actions workflow
* It looks like the badge on the README.md is for the fork [owhere/bricsauthenticator], not this repository

I don't think we should hold off merging for these issues, though. Instead I think we should add an issue to the repo so we can follow up on this after merging the PR.

jcwomack committed:
Remove unneeded `make test` from Actions workflow

`pytest` with coverage runs all the tests, so we only need `make coverage`

jcwomack commented:
Just sneaked in a small change to the Actions workflow. It turned out we were running the tests twice (once without coverage, once with coverage). I removed the step that runs without coverage, since this was a waste of Actions minutes.

jcwomack commented:
I also added a new issue isambard-sc/jupyter-platform#78 to suggest that we should add further tests which check behaviour in the case of realistic valid/invalid JWTs. Again, I don't think this should hold up merging this particular PR.
